src/ui/api/metadata.go

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/vmware/harbor/src/common/models"
+	"github.com/vmware/harbor/src/common/utils/log"
+	"github.com/vmware/harbor/src/ui/promgr/metamgr"
+)
+
+// MetadataAPI ...
+type MetadataAPI struct {
+	BaseController
+	metaMgr metamgr.ProjectMetadataManager
+	project *models.Project
+	name    string
+}
+
+// Prepare ...
+func (m *MetadataAPI) Prepare() {
+	m.BaseController.Prepare()
+
+	m.metaMgr = m.ProjectMgr.GetMetadataManager()
+
+	// the project manager doesn't use a project metadata manager
+	if m.metaMgr == nil {
+		log.Debug("the project manager doesn't use a project metadata manager")
+		m.RenderError(http.StatusMethodNotAllowed, "")
+		return
+	}
+
+	id, err := m.GetInt64FromPath(":id")
+	if err != nil || id <= 0 {
+		text := "invalid project ID: "
+		if err != nil {
+			text += err.Error()
+		} else {
+			text += fmt.Sprintf("%d", id)
+		}
+		m.HandleBadRequest(text)
+		return
+	}
+
+	project, err := m.ProjectMgr.Get(id)
+	if err != nil {
+		m.ParseAndHandleError(fmt.Sprintf("failed to get project %d", id), err)
+		return
+	}
+
+	if project == nil {
+		m.HandleNotFound(fmt.Sprintf("project %d not found", id))
+		return
+	}
+
+	m.project = project
+
+	switch m.Ctx.Request.Method {
```

Comment:
Why not use if

if !(m.Ctx.Request.Method == http.MethodGet && project.IsPublic){
  //do auth
}

if m.Ctx.Request.Method == other {
  //not allowed
}